Mathlib/Tactic/Simproc/FinsetInterval.lean

```diff
@@ -0,0 +1,300 @@
+
+/-!
+# Simproc for intervals of natural numbers
+-/
+
+open Qq Lean Finset
+
+namespace Mathlib.Tactic.Simp
+namespace Nat
+variable {m n : ℕ} {s : Finset ℕ}
+
+private lemma Icc_eq_empty_of_lt (hnm : n.blt m) : Icc m n = ∅ := by simpa using hnm
+
+private lemma Icc_eq_insert_of_Icc_succ_eq (hmn : m.ble n) (hs : Icc (m + 1) n = s) :
+    Icc m n = insert m s := by rw [← hs, insert_Icc_add_one_left_eq_Icc (by simpa using hmn)]
+
+private lemma Ico_succ_eq_of_Icc_eq (hs : Icc m n = s) : Ico m (n + 1) = s := by
+  rw [← hs, Ico_add_one_right_eq_Icc]
+
+private lemma Ico_zero (m : ℕ) : Ico m 0 = ∅ := by simp
+
+private lemma Ioc_eq_of_Icc_succ_eq (hs : Icc (m + 1) n = s) : Ioc m n = s := by
+  rw [← hs, Icc_add_one_left_eq_Ioc]
+
+private lemma Ioo_eq_of_Icc_succ_pred_eq (hs : Icc (m + 1) (n - 1) = s) : Ioo m n = s := by
+  rw [← hs, ← Icc_add_one_sub_one_eq_Ioo]
+
+private lemma Iic_eq_of_Icc_zero_eq (hs : Icc 0 n = s) : Iic n = s := hs
+
+private lemma Iio_succ_eq_of_Icc_zero_eq (hs : Icc 0 n = s) : Iio (n + 1) = s := by
+  erw [Iio_eq_Ico, Ico_add_one_right_eq_Icc, hs]
+
+private lemma Iio_zero : Iio 0 = ∅ := by simp
+
+end Nat
+
+namespace Int
+variable {m n : ℤ} {s : Finset ℤ}
+
+private lemma Icc_eq_empty_of_lt (hnm : n < m) : Icc m n = ∅ := by simpa using hnm
+
+private lemma Icc_eq_insert_of_Icc_succ_eq (hmn : m ≤ n) (hs : Icc (m + 1) n = s) :
+    Icc m n = insert m s := by rw [← hs, insert_Icc_add_one_left_eq_Icc (by simpa using hmn)]
+
+private lemma Ico_eq_of_Icc_pred_eq (hs : Icc m (n - 1) = s) : Ico m n = s := by
+  rw [← hs, Icc_sub_one_right_eq_Ico]
+
+private lemma Ioc_eq_of_Icc_succ_eq (hs : Icc (m + 1) n = s) : Ioc m n = s := by
+  rw [← hs, Icc_add_one_left_eq_Ioc]
+
+private lemma Ioo_eq_of_Icc_succ_pred_eq (hs : Icc (m + 1) (n - 1) = s) : Ioo m n = s := by
+  rw [← hs, ← Icc_add_one_sub_one_eq_Ioo]
+
+private lemma Iio_zero : Iio 0 = ∅ := by simp
+
+end Int
+
+/-- Given natural numbers `m` and `n` and corresponding natural literals `em` and `en`,
+returns `(s, ⊢ Finset.Icc m n = s)`. -/
+def evalFinsetIccNat (m n : ℕ) (em en : Q(ℕ)) :
+    MetaM ((s : Q(Finset ℕ)) × Q(.Icc $em $en = $s)) := do
+  -- If `m = n`, then `Icc m n = {m}`. We handle this case separately because `insert m ∅` is
+  -- not syntactically `{m}`.
+  if m = n then
+    have : $em =Q $en := ⟨⟩
+    return ⟨q({$em}), q(Icc_self _)⟩
+  -- If `m < n`, then `Icc m n = insert m (Icc (m + 1) n)`.
+  else if m < n then
+    let hmn : Q(Nat.ble $em $en = true) := (q(Eq.refl true) :)
+    have em' : Q(ℕ) := mkNatLitQ (m + 1)
+    have : $em' =Q $em + 1 := ⟨⟩
+    let ⟨s, hs⟩ ← evalFinsetIccNat (m + 1) n em' en
+    return ⟨q(insert $em $s), q(Nat.Icc_eq_insert_of_Icc_succ_eq $hmn $hs)⟩
+  -- Else `n < m` and `Icc m n = ∅`.
+  else
+    let hnm : Q(Nat.blt $en $em = true) := (q(Eq.refl true) :)
+    return ⟨q(∅), q(Nat.Icc_eq_empty_of_lt $hnm)⟩
+
+/-- Given integers `m` and `n` and corresponding integer literals `em` and `en`,
+returns `(s, ⊢ Finset.Icc m n = s)`. -/
```

Comment:
```suggestion
returns `(s, ⊢ Finset.Icc m n = s)`.

This cannot be easily merged with `evalFinsetIccNat` since they require different
handling of numerals for `ℕ` and `ℤ`.
-/
```